src/System.Linq/tests/EnumerableTests.cs

```diff
@@ -0,0 +1,237 @@
+            Assert.Equal(1.0/3, new Queue<int>(oneByThree).Average());
+        }
+
+        [Fact]
+        public void All()
+        {
+            var array = Enumerable.Range(1, 10).ToArray();
+            Assert.True(array.All(i => i > 0));
+            for (var j = 1; j <= 10; j++)
+                Assert.False(array.All(i => i > j));
+        }
+
+        [Fact]
+        public void Any()
+        {
+            var array = Enumerable.Range(1, 10).ToArray();
+            for (var j = 0; j <= 9; j++)
+                Assert.True(array.Any(i => i > j));
+            Assert.False(array.Any(i => i > 10));
+        }
+
+        private void FindCountAndValidate<T>(IEnumerable<T> enumerable, int expectedCount)
+        {
+            Assert.Equal(expectedCount, enumerable.Count());
+            Assert.Equal(expectedCount, enumerable.ToList().Count());
+            Assert.Equal(expectedCount, enumerable.ToArray().Count());
+            Assert.Equal(expectedCount, new Stack<T>(enumerable).Count());
+            Assert.Equal(expectedCount, new Queue<T>(enumerable).Count());
+        }
+
+        [Fact]
+        public void Count()
+        {
+            const int count = 100;
+            var range = Enumerable.Range(1, count).ToArray();
+            FindCountAndValidate(range, count);
+            FindCountAndValidate(range.Select<int, long>(i => i), count);
+            FindCountAndValidate(range.Select<int, float>(i => i), count);
+            FindCountAndValidate(range.Select<int, double>(i => i), count);
+            FindCountAndValidate(range.Select<int, decimal>(i => i), count);
+        }
+
+        private void FindDistinctAndValidate<T>(IEnumerable<T> original)
+        {
+            // Convert to list to avoid repeated enumerations of the enumerables.
+            var originalList = original.ToList();
+            var distinctList = originalList.Distinct().ToList();
+
+            // Ensure the result doesn't contain duplicates; sort it first to make this easier.
+            // Notice that we assume the correctness of OrderBy() here, which is tested
+            // in another unit.
+            for (var i = 0; i < distinctList.Count - 1; i++)
+            {
+                Assert.False(distinctList[i].Equals(distinctList[i + 1]),
+                    "There are some duplicates in the result of Distinct().");
+            }
+
+            // Ensure that every element in 'array' exists in 'distinctList'.
+            foreach (var i in originalList)
+            {
+                var correct = false;
+                foreach (var j in distinctList)
+                {
+                    if (i.Equals(j))
+                        correct = true;
+                }
+                Assert.True(correct, "Some elements in the original enumerable are missing in the result of Distinct().");
+            }
+
+            // Ensure that every element in 'distinctList' exists in 'originalList'.
+            foreach (var i in distinctList)
```

Comment:
149 +            // Ensure that every element in 'array' exists in 'distinctList'. 
    150 +            foreach (var i in originalList) 
    151 +            { 
    152 +                var correct = false; 
    153 +                foreach (var j in distinctList) 
    154 +                { 
    155 +                    if (i.Equals(j)) 
    156 +                        correct = true; 
    157 +                } 
    158 +                Assert.True(correct, "Some elements in the original enumerable are missing in the result of Distinct()."); 
    159 +            } 

Yeah, no, lets not do tests this way please.  I've been cleaning up this sort of thing from the PLINQ tests.

You should be using:

    foreach(var i in originalList)
    {
        Assert.Contains(i, distinctList);
    }

Or even:

    Assert.All(originalList, i => Assert.Contains(i, distinctList));

...which will even print out the index of the offending element, although it means the output wouldn't be pipelined (because internally it calls `ToArray()`, although you shouldn't rely on that particular behavior)...
You are planning on pipelined vs non-pipelined output tests, aren't you?

Side note - despite the comments in that test, you _aren't_ sorting it first (it assumes sorted input).  It would be better to use some sort of set for this verification (either grab `HashSet`, or you can make a trivial helper set class in about 10 lines).